README.md
[![pypi](https://img.shields.io/pypi/v/pysigfox-client.svg)](https://pypi.org/project/pysigfox-client/)
[![Build Status](https://travis-ci.org/optimdata/pysigfox.svg?branch=master)](https://travis-ci.org/optimdata/pysigfox)
[![codecov](https://codecov.io/gh/optimdata/pysigfox/branch/master/graph/badge.svg)](https://codecov.io/gh/optimdata/pysigfox)
[![Documentation Status](https://readthedocs.com/projects/inuse-pysigfox/badge/?version=latest&token=9ae214b02c982eb9632fb078aabb9b47db229adae27be4357f4d0b381e8f0cff)](https://inuse-pysigfox.readthedocs-hosted.com/en/latest/?badge=latest)

# pysigfox

`pysigfox` is an unofficial client for the V2 [sigfox API](https://support.sigfox.com/apidocs). 

Inspired from [python-sigfox-backend-api](https://github.com/mjuenema/python-sigfox-backend-api).

## Resources

- [Official sigfox API documentation](https://support.sigfox.com/apidocs)
- [Documentation](https://inuse-pysigfox.readthedocs-hosted.com/en/latest/)


HISTORY.md
## History

### 0.1.0

- First version

### 0.1.1

- Fix cast to integer for datetime.datetime to timestamps in milliseconds

### 0.1.2

- Add constants for device end points

### 0.1.3

- Add api user end point

### 0.1.4

- Add global timeout


tests/test_pysigfox.py
# -*- coding: utf-8 -*-
import datetime
import re
import os
import sys

import pytest
import requests_mock

BASE_DIRECTORY = os.path.join(os.path.dirname(__file__), "..")  # NOQA
sys.path.insert(0, BASE_DIRECTORY)  # NOQA

from pysigfox import (
    Sigfox,
    SigfoxTooManyRequestsError,
    SigfoxBadStatusError,
    from_ms_timestamp,
    to_ms_timestamp,
    to_bytearray,
    constants,
)

DEVICES = {
    "data": [
        {
            "id": "device_1",
            "name": "device_1",
            "lastCom": 1_559_802_207_000,
            "state": 0,
            "comState": 5,
            "deviceType": {"id": "1"},
            "group": {"id": "1"},
            "lqi": 4,
        }
    ],
    "paging": {},
}

DEVICE_MESSAGES = {
    "data": [
        {
            "device": {"id": "device_1"},
            "time": 1_613_984_321_000,
            "data": "010ea00cc764",
            "lqi": 2,
        },
        {
            "device": {"id": "device_1"},
            "time": 1_613_897_925_000,
            "data": "010e320ca664",
            "lqi": 2,
        },
    ],
    "paging": {},
}

USER = {
    "id": "test",
    "group": {"id": "1", "name": "group"},
    "name": "test_name",
    "timezone": "UTC",
    "creationTime": 1_613_489_692_898,
}


# sigfox API mocker and pollux mocker
class SigfoxMocker(requests_mock.mock):
    def __init__(self, status_code=200, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.raise_error = status_code
        self.call_nb = 0

        def device_messages(request, context):
            if self.call_nb == 0:
                self.call_nb += 1
                results = DEVICE_MESSAGES.copy()
                results["paging"] = {"next": request.url + "?page=2"}
                return results
            return DEVICE_MESSAGES

        # devices
        self.get(
            "https://api.sigfox.com/v2/devices", json=DEVICES, status_code=status_code
        )

        # device messages
        message_matcher = re.compile(
            r"^(https://api\.sigfox\.com/v2/devices/)(.*)(/messages)(.*)$"
        )
        self.get(message_matcher, json=device_messages, status_code=status_code)

        # device
        device_matcher = re.compile(r"^(https://api\.sigfox\.com/v2/devices/)(\w+)$")
        self.get(device_matcher, json={}, status_code=status_code)

        # user
        user_matcher = re.compile(r"^(https://api\.sigfox\.com/v2/api-users/)(\w+)$")
        self.get(user_matcher, json=USER, status_code=status_code)


def test_sigfox():
    client = Sigfox(username="test", password="test")
    print(client)
    with SigfoxMocker():
        assert len(client.devices(list_all=True)) == 1
        assert len(client.device_messages("device_1", list_all=True)) == 4
        assert client.device("device_1") == {}
        assert client.api_user()["name"] == "test_name"
        assert client.api_user("test")["name"] == "test_name"
    with SigfoxMocker():
        assert len(client.device_messages("device_1")) == 2

    with SigfoxMocker(status_code=429):
        with pytest.raises(SigfoxTooManyRequestsError):
            client.devices()

    with SigfoxMocker(status_code=404):
        with pytest.raises(SigfoxBadStatusError):
            client.devices()

    with requests_mock.mock() as mock:
        mock.get(
            "https://api.sigfox.com/v2/devices/device_1/messages",
            json={
                "data": [],
                "paging": {
                    "next": "https://api.sigfox.com/v2/devices/device_1/messages?page=2"
                },
            },
        )
        client.request("GET", "devices/device_1/messages")

    with requests_mock.mock() as mock:
        mock.get(
            "https://api.sigfox.com/v2/devices/device_1/messages",
            json={
                "data": [],
                "paging": {
                    "next": "https://api.sigfox.com/v2/devices/device_1/messages?"
                },
            },
        )
        client.request("GET", "devices/device_1/messages")


def test_utils():
    timestamp = 1_613_984_321_000
    dt = from_ms_timestamp(timestamp)
    my_dt = datetime.datetime(year=2021, month=2, day=22, hour=8, minute=58, second=41)
    assert dt == my_dt
    assert str(to_ms_timestamp(my_dt)) == str(timestamp)

    dt = from_ms_timestamp(timestamp)
    dt = dt.replace(tzinfo=datetime.timezone(datetime.timedelta(hours=4)))
    timestamp = to_ms_timestamp(dt)

    my_dt = datetime.datetime(
        year=2021,
        month=2,
        day=22,
        hour=8,
        minute=58,
        second=41,
        tzinfo=datetime.timezone(datetime.timedelta(hours=4)),
    )
    assert dt == my_dt
    assert str(to_ms_timestamp(my_dt)) == str(timestamp)

    value = "010ea00cc764"
    assert to_bytearray(value).hex() == value

    assert to_ms_timestamp(None) is None
    assert from_ms_timestamp(None) is None
    assert to_bytearray(None) is None


def test_constants():
    assert len(constants.LQIS) == 5
    assert len(constants.COM_STATES) == 6
    assert len(constants.STATES) == 8


pysigfox/sigfox.py
import functools
import json
import logging
import time

import requests
import urllib.parse

from .exceptions import (
    SigfoxConnectionError,
    SigfoxBadStatusError,
    SigfoxResponseError,
    SigfoxTooManyRequestsError,
)
from .utils import to_ms_timestamp


# The rate limit for the device messages end point (unit: query/seconds)
RATE_LIMIT_DEVICE_MESSAGES = 1

logger = logging.getLogger("client")


class Sigfox(object):
    """
    `Sigfox V2 API client <https://support.sigfox.com/apidocs>`_

    Inspired from https://github.com/mjuenema/python-sigfox-backend-api
    """

    def next(self, *args, **kwargs):
        """Fetch the next page of results for some methods.

           Call this method whenever another method has returned only
           a subset of the results.

           >>> messages = s.device_messages('4d3091a05ee16b3cc86699ab')
           >>> while s.next:
           ...     messages += s.next()
           >>> len(messages)
           310
           >>> devices = s.devices()
           >>> while s.next:
           ...     devices += s.next()
           >>> len(devices)
           22

           .. warning:: Be mindful that this may return a huge number of
                        results if used exactly as in the example above.

        """

        # `Sigfox.next()` will be set in `Sigfox.request()` to ``None`` or
        # a `functool.partial(...)` method matching the original method.
        # The code here is only for documentation purposes.
        pass  # pragma: no cover

    def __init__(self, username, password, timeout=None):
        self.account = username
        self.username = username
        self.password = password
        self.timeout = timeout
        self.session = requests.Session()
        self.session.auth = (username, password)
        self.base_url = "https://api.sigfox.com/v2/"
        self.client_type = "Sigfox"
        self.next = None

    def _build_url(self, url):
        return self.base_url + url

    def request(self, method, route, params=None, **kwargs):
        """
        Base request method that can be used if a route is not configured in this client.
        This method handles the pagination and perform the query using a `request.request <https://requests.readthedocs.io/en/latest/api/#requests.request>`_

        It returns the `data` stored in the response and update the `next` method if there is a pagination.

        """
        url = self._build_url(route)
        kwargs.setdefault("timeout", self.timeout)
        try:
            response = self.session.request(
                method=method, url=url, params=params, **kwargs
            )
        except requests.exceptions.ConnectionError as e:  # pragma: nocover
            raise SigfoxConnectionError(str(e))  # pragma: nocover

        # HTTP 429: too many requests
        if response.status_code == 429:
            raise SigfoxTooManyRequestsError()

        if response.status_code != 200:  # pragma: nocover
            raise SigfoxBadStatusError(
                f"Bad status {response.status_code} : {response.text}"
            )

        try:
            response = response.json()
        except json.decoder.JSONDecodeError:  # pragma: nocover
            raise SigfoxResponseError(
                "Cannot deserialize json from %s" % response.content
            )

        data = response.get("data", response)

        # Set `Sigfox.next()` by extracting the parameters from the 'next' URL and
        # currying the self.request().
        self.next = None
        if "paging" in response and "next" in response["paging"]:
            next_params = dict(
                urllib.parse.parse_qsl(response["paging"]["next"].split("?")[1])
            )
            if next_params:
                if params:
                    params.update(next_params)
                else:
                    params = next_params
                self.next = functools.partial(
                    self.request, method, route, params, **kwargs
                )
            else:
                self.next = None

        return data

    def devices(self, list_all=False, **kwargs):
        """
        `Retrieve a list of devices according to visibility permissions and request filters. <https://support.sigfox.com/apidocs#operation/listDevices>`_

        :param bool list_all: extract all devices (this endpoint is paginated).
        """
        params = kwargs.pop("params", {})
        results = self.request("GET", "devices", params=params, **kwargs)
        if list_all:
            self._handle_next_pages(results)

        return results

    def device(self, device_code, **kwargs):
        """
        `Retrieve information about a given device <https://support.sigfox.com/apidocs#operation/getDevice>`_

        :param str device_code: The device identifier.
        """
        return self.request("GET", f"devices/{device_code}", **kwargs)

    def api_user(self, api_user_id=None, **kwargs):
        """
        `Retrieve information about a given API user. <https://support.sigfox.com/apidocs#operation/getApiUser>`_

        :param str api_user_id: The API user identifier. If none we use the api_user_id used for this client.
        """
        if api_user_id is None:
            api_user_id = self.username
        return self.request("GET", f"api-users/{api_user_id}", **kwargs)

    def device_messages(
        self, device_code, list_all=False, since=None, before=None, **kwargs
    ):
        """
        `Retrieve a list of messages for a given device according to request filters, with a 3-day history. <https://support.sigfox.com/apidocs#operation/getDeviceMessagesListForDevice>`_
        Messages are retrieved from the newest to the oldest.


        :param str device_code: The device identifier.
        :param bool list_all: If ``list_all``, extract all the devices (this endpoint is paginated). This api has a `rate limiting <https://support.sigfox.com/docs/api-rate-limiting>`_ of 1 request per second.
        :param datetime.datetime since: the starting date time of the period to fetch. can be null.
        :param datetime.datetime before: the ending date time of the period to fetch. can be null.
        """
        # process period
        since = to_ms_timestamp(since)
        before = to_ms_timestamp(before)
        params = kwargs.pop("params", {})
        params.update({"since": since, "before": before})
        results = self.request(
            "GET", f"devices/{device_code}/messages", params=params, **kwargs
        )
        if list_all:
            self._handle_next_pages(results, rate_limit=RATE_LIMIT_DEVICE_MESSAGES)
        return results

    def _handle_next_pages(self, results, rate_limit=None):
        """
        Handle pagination of some end point by making sure we retrieve all the data
        (even if the endpoint has a rate limitation)

        :param list results: The list where to append the paginated results.
        :param int rate_limit: The rate limit (duration in seconds between two calls for this end point).
        """
        while self.next:
            try:
                results += self.next()
            except SigfoxTooManyRequestsError:  # pragma: no cover
                logger.debug(
                    f"Too many requests on the API. Sleep for {rate_limit} seconds."
                )
                time.sleep(rate_limit)
                continue
        return results
